### ISSUE TITLE
enforce-switch-style: an option to disable the rule for `switch` with only expression cases

### ISSUE THREAD

**Issue by gannett-ggreer** [mgechev/revive#1434]
**Is your feature request related to a problem? Please describe.**
When a `switch` has only expressions it is more similar to a bunch of `if { } else if { }` blocks and forcing a `default:` for those is like forcing an `else { }` block at the end of an `if` chain, even if one isn't necessary.

**Describe the solution you'd like**
The `enforce-switch-style` lint check should never trigger from expression-only switches.

**Describe alternatives you've considered**
Whether the linter triggers on expression-only switches or not could be configurable, if some scenario such a warning is valuable.

**Additional context**
An "expression-only switch" is code like:
```
  Error: authoring/image.go:186:2: enforce-switch-style: switch must have a default case clause (revive)
  	switch {
  	case width > height && width > maxWidth:
  		height *= maxWidth / width
  		width = maxWidth
  		if height > maxHeight {
  			width *= maxHeight / height
  			height = maxHeight
  		}
  	case height > width && height > maxHeight:
  		width *= maxHeight / height
  		height = maxHeight
  	case height == width && width > maxHeight:
  		width = maxWidth
  		height = maxHeight
  	}```

**Comment by chavacava** [comment]
Hi @gannett-ggreer thanks for reporting the issue. 

Forcing a `default` (or a final `else` in an `if-then-else` chain) can be subject of endless discussions (i.e.; it's a very opinionated subject)

Enforcing mandatory `default` or `else` is the norm in other languages with these control flow instructions (like [Java](https://rules.sonarsource.com/java/RSPEC-126/))

I think that not enforcing `default` for "expression-only" switches will  make the rule less useful.

I (try to) add a default case to my switches. When dealing with cases where all expression are handled I add a comment to the default case, something like:

```go
switch {
  case a > b :
    // ...
  case a < b :
    // ...
  default: // a == b
    // ...
}
```

**Comment by ccoVeille** [comment]
I feel like when someone is facing an issue reported with enforce-switch-style, there are three possibilities:

- the rule doesn't match the codebase way to do, and so almost all switch would be reported

Then disabling the rule might be a good option

- the rule reports something OK in most of the cases

Here you report the expression-only case, but for other switches it's OK

 then the rule could be disabled for this switch or an empty default could be added.

- it's legitimate to report a default seems missing.

Personal note, I have a lot of switch in my codebase, only a very few of them are let's call them "enum-only" case, most of them are using the expression-only case. We are using a default in almost all of them. I added a ignore rule when a default was not necessary.

**Comment by gannett-ggreer** [comment]
[ccoVeille](https://github.com/ccoVeille)

> I feel like when someone is facing an issue reported with enforce-switch-style, there are three possibilities:
> 
>     * the rule doesn't match the codebase way to do, and so almost all switch would be reported
> 
> 
> Then disabling the rule might be a good option
> 
>     * the rule reports something OK in most of the cases
> 
> 
> Here you report the expression-only case, but for other switches it's OK
> 
> then the rule could be disabled for this switch or an empty default could be added.
> 
>     * it's legitimate to report a default seems missing.
> 
> 
> Personal note, I have a lot of switch in my codebase, only a very few of them are let's call them "enum-only" case, most of them are using the expression-only case. We are using a default in almost all of them. I added a ignore rule when a default was not necessary.

From our experience on code bases with this being enabled when it wasn't before, almost all of the `switch var {}` cases having a `default:` made sense because there was an enumerable list of possibilities and it was useful to differentiate between accidentally omitting a value or intentionally limiting the values.  For cases where we omitted values intentionally we added `default: // some comment why other values aren't present`.

But for all of the `switch {}` cases it more was more like an `if {} else if {}` chain and none of those made sense to have a default because they weren't trying to be exhaustive, nor would there be any value in adding a `default: // comment` why because the comment wouldn't add any value by saying "no other conditions matter" or something to that effect every time.  That's why I likened it to forcing an `else {}` block on every `if {}`, which is syntactically and semantically useless.

Currently disabling is an all-or-nothing affair so at least being able to limiting it to value switches is a benefit, even if I can't see a value to having it on for expressions-only switch. Or possibly give both cases an entirely different linter name rather than configuring the existing one, so one can be disabled and the other enabled.

**Comment by ccoVeille** [comment]
I suggested something there

- https://github.com/mgechev/revive/issues/1438#issuecomment-3121346521
